[RM Ch07] Software supply chain security (#2913)
cntt-n/CNTT: doc/ref_model/chapters/chapter07.rst

organisations are experiencing these threats. Software suppliers can be targeted and by rebound effect their 
customers, as seen with the backdoors affecting SolarWinds products and detailed in
`Defending against SolarWinds attacks 
<https://www.techtarget.com/searchsecurity/news/252494495/Defending-against-SolarWinds-attacks-What-can-be-done>`_.
Open-source code weaknesses can also be utilised by attackers, the Log4J vulnerability, impacting lots of 
applications, is a recent example in this field. The cyber security of software supply chain are often not 
taken into account. Gouvernments are alerting and requesting actions to face these risks. The British 
government is hardening the law and standards of cyber security for the supply chain. The US government 
requested actions to enhance the software supply chain security. The security of software supply chain 
is a challenge pointed out by the European Network and Information Security Agency, ENISA, 
in the report `NFV Security in 5G - Challenges and Best Practices 
<https://www.enisa.europa.eu/publications/nfv-security-in-5g-challenges-and-best-practices>`_.
 

Software security
~~~~~~~~~~~~~~~~~

> Software supply chain safety is crucial and can be a complex task in virtualised and containerised environments

Reviewer: Should be security, not safety.

Author: @ulikleber I think that both work

Reviewer 2: @karinesevilla , @ulikleber: this is a very subtle linguistic difference, and none of us was born in an English speaking country (I guess...).  I would refer us to an interesting discussion in  http://www.differencebetween.net/language/words-language/difference-between-safety-and-security/.  My  feeling from reading the comparison there, that "security" would be more appropriate in our case.

Reviewer 3: May I suggest a compromise and we use "security and safety"?

Reviewer: I think using safety when talking about software vulnerability is really strange. 
Safety in the context of computers is for instance used, when you program a robot that when moving around could hurt people. So the software reaction times and reliability are critical.

Author: Thanks all for sharing your opinions. There is a consensus in favor of "security". I modified the text.

Author: Thanks for sharing your opinions. There is a consensus to use the word security, I made the modification

Software Supply Chain Security
------------------------------

> Software supply chain attacks are increasing worldwilde and can cause serious damages. Many entreprises and 

Reviewer 3: Do we have a reference to some incidents or data?

Author: References added into text

> organisations are experiencing these threats. Software suppliers can be targeted and by rebound effect their 

Reviewer 3: "rebound"?
```suggestion
organisations are experiencing these threats. Software suppliers can be targeted, and by rebound affect their 
```

Author: @pgoyal01 in this sentence "Rebound effect" can not be replaced by affect. I can change to : and by rebound their customers can be affected. it would works?

Author: Text changed, please Check if Okay

customers, as seen with the backdoors affecting SolarWinds products and detailed in
`Defending against SolarWinds attacks 
<https://www.techtarget.com/searchsecurity/news/252494495/Defending-against-SolarWinds-attacks-What-can-be-done>`_.

> Open-source code weaknesses can also be utilised by attackers, the Log4J vulnerability, impacting lots of 

Reviewer 3: ```suggestion
Open-source code weaknesses can also be utilised by attackers, the `Log4J <https://www.cisa.gov/uscert/apache-log4j-vulnerability-guidance>`_
vulnerability, impacting many 
```

Author: Thanks Pankaj for this pointer! Suggestion added

> applications, is a recent example in this field. The cyber security of software supply chain are often not 

Reviewer 3: ```suggestion
applications, is a recent example in this field. When addressing cyber security vulnerabilities, 
the vulnerabilities  of the software supply chain are often not 
```

Author: Thanks, modified

> taken into account. Governments are alerting and requesting actions to face these risks. The British 

Reviewer 3: ```suggestion
taken into account. Some governments are already alerting and requesting actions to face these risks. The British 
```

Author: Thanks , added

government is hardening the law and standards of cyber security for the supply chain. The US government 

> requested actions to enhance the software supply chain security. The security of software supply chain 

Reviewer 3: ```suggestion
requested actions to enhance the software supply chain security. The security of the software supply chain 
```

Author: Thanks, added

> is a challenge pointed out by the European Network and Information Security Agency, ENISA, 

Reviewer 3: ```suggestion
is also a challenge pointed out by the European Network and Information Security Agency, ENISA, 
```

Author: Thanks, added

in the report `NFV Security in 5G - Challenges and Best Practices 
<https://www.enisa.europa.eu/publications/nfv-security-in-5g-challenges-and-best-practices>`_.
 

Software security
~~~~~~~~~~~~~~~~~

Software supply chain safety is crucial and can be a complex task in virtualised and containerised environments

> with a greater attack surface. All software components must be trusted, from commercial software, open-source 

Reviewer 3: The sentence, "Software supply chain safety is crucial and can be a complex task in virtualised and containerised environments with a greater attack surface", can be read as that this is only an issue where the environments have a "greater attack surface". 

One suggestion:
"Software supply chain safety is crucial and is made complex by the greater attack surface provided by the many different supply chains in virtualised, containerised, and edge environments."

Author: Thanks, It is better, added

> code to proprietary software. The SAFECode paper “Managing Security Risks Inherent in the Use of Third-party 

Reviewer 2: Should we also mention security of integration of these components?  Or this is not a subject of Supply Chain Security?

Author: Yes, It is. Addition in text

> Components” provides a detailed risk management approach.

Reviewer 3: Reference to the paper

Author: reference added

To secure software code, the following methods must be applied:

> -  Use best practices coding such as design pattern recommended in the `Twelve-Factor App <https://12factor.net/>`_ or `OWASP “Secure Coding Practices - Quick Reference Guide” <https://owasp.org/>`_

Reviewer 3: how do these best practices help with supply chain security?

Also, can the bullets be re-ordered?

Author: List re-ordered

-  Require suppliers to provide a Software Bill of Materials to identify all the components part of their product’s software releases with their dependencies, and eventually identify the open source modules
-  Use trusted, authenticated and identified software images that are provided by authenticated software distribution portals

> -  Do threat modelling, as described in the document “Tactical Threat Modeling” published by SAFECode

Reviewer 2: Can you provide a reference to this document “Tactical Threat Modeling”?

Author: Reference added

> -  Test the software in a pre-production environment to validate integration

Reviewer 2: So integration comes into picture after all! Shouldn't be "to validate security of integration"?

Author: I added the bullet: Secure the integration process by securing the software production pipeline

-  Actively monitor the open source software repositories to determine if new versions have been released that address identified vulnerabilities discovered in the community
-  Report and remove vulnerabilities by upgrading components using authenticated software update distribution portals
-  Adopt a DevSecOps approach and rely on testing automation throughout the software build, integration, delivery, deployment, and runtime operation to perform automatic security check, as described in section ”Infrastructure as a Code Security”

Open-Source Software Security
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Open-source code is present in Cloud Infrastructure software from BIOS, host Operating System to virtualisation layer 
components, the most obvious being represented by Linux, KVM, QEMU, OpenStack, and Kubernetes. Workloads components 
can also be composed of open source code. The proportion of open-source code to an application source code can vary. 
It can be partial or total, visible or not. Open-source code can be upstream code coming directly from open-source 

> public repositories or code within a commercial application or network function. 

Reviewer 2: Often an open-source software customised by an organisation (telecom operator in our case) is also understood as "open-source".  I am in two minds about it but I think we should specify whether we consider this as an open-source or a proprietary source based on open-source (I would vote for the latter...)

Reviewer 2: On the second thought, if we consider "code within a commercial application or network function" as open source (because of its open source) then perhaps we should include also an open-source code within "a bespoke application or network function"?

Author: I vote also for  "proprietary source based on open-source" in this case.

> The strength of open-source code is the availability of code source developed by a community which maintain and improve it. 

Reviewer 4: minor fix

```suggestion
The strength of open-source code is the availability of code source developed by a community which maintains and improves it. 
```

Author: Thanks @gkunz , modified

Open-source code integration with application source code helps to develop and produce applications faster. But, in return, 
it can introduce security risks if a risk management DevSecOps approach is not implemented. The GSMA white paper, “Open 

> Networking & the Security of Open Source Software Deployment - Future Networks”, alerts on these risks and addresses the 

Reviewer 2: Reference please

Author: Reference added

> challenges coming with open-source code usage. Amongst these risks for security, we can mention a poor quality code 

Reviewer 4: ```suggestion
challenges coming with open-source code usage. Amongst these risks for security, we can mention a poor code quality
```

Author: Thanks, changed

containing security flaws, an obsolete code with known vulnerabilities, and the lack of knowledge of open source 
communities’ branches activity. An active branch will come with bugs fixes, it will not be the case with an inactive branch. 
The GSMA white paper develops means to mitigate these security issues.

Poor code quality is a factor of risk. Open-source code advantage is its transparency, code can be inspected by tools with 
various capabilities such as open-source software discovery and static and dynamic code analysis.

Each actor in the whole chain of software production must use a dedicated internal isolated repository separated from the 
production environment to store vetted open-source content, which can include images, but also installer and utilities. 
These software packages must be signed and the signature verified prior to packages or images installation. 
Access to the repository must be granted by a dedicated authorization. The code must be inspected and vulnerabilities 

> identified as described previously. After validating the software is risk free, it can be moved to the appropriate production 

Reviewer 4: Nothing is ever free of risks ;-)

```suggestion
identified as described previously. After validating the software, it can be moved to the appropriate production 
```

Author: Right, modified

repository.

SBOM
~~~~

It is crucial to identify the software components and their origins. The Software Bill of Materials (SBOM), 
described by `US NTIA <https://www.ntia.gov/SBOM>`__ (National Telecommunications and Information Administration), 
is an efficient tool to identify software components. The SBOM is an inventory of software components and the relationships 
between them. NTIA describes how to establish an SBOM and provides SBOM standard data formats. In case of vulnerability 
detected for a component, the SBOM inventory is an effective means to identify the impacted component and provide remediation.

A transparent software supply chain offers benefits for vulnerabilities remediation, but also for licensing management and it 
provides assurance of the source and integrity of components. In order to use the SBOM efficiently and spread its adoption, 
information must be generated and shared in a standard format. This format must be machine-readable to allow automation. The 

> NTIA white paper `“Survey of existing SBOM formats and standards” <https://www.ntia.doc.gov/files/ntia/publications

Reviewer 4: How about referecing the "The Minimum Elements for an SBOM" NTIA paper, which is newer and lists CycloneDX as well? I have the impression, the current favorites in the industry are SPDX, CycloneDX (in this order) and SWID is not often discussed. Please correct me, if this impression is wrong.

Author: @gkunz Yes, totally agree. The plan is to restructure the section and then to develop the content on SBOM. Another PR associated with issue#2793 will be created for SBOM when this one is merged.

/ntia_sbom_formats_and_standards_whitepaper_-_version_20191025.pdf>`__ identifies two key formats covering baseline SBOM 

> information: Software Package Data eXchange (SPX) and Software Identification Tags (SWID tags).

Reviewer 2: Should be SPDX, not SPX

Author: Yes thanks, it is a typo, now modified

publishers and has been published as the standard ISO/IEC 19770-2. The specification defines four types of SWID tags: primary, 
patch, corpus, and supplemental to describe a software component.

For both formats, SPDX and SWID Tags, tools are available.

Vulnerability identification
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Vulnerability management must be continuous: from development to runtime, not only on the development process, but during all 
the life of the application or workload or service. When a public vulnerability on a component is released, the update of the 
component must be triggered. When an SBOM recording the code composition is provided, the affected components will be easier 
to identify. It is essential to remediate the affected components as soon as possible, because the vulnerability can be 
exploited by attackers who can take the benefit of code weakness.

The CVE and the CVSS must be used to identify vulnerabilities and their severity rating. The CVE identifies, defines, and 
catalogues publicly disclosed cybersecurity vulnerabilities while the CVSS is an open framework to calculate the 
vulnerabilities' severity score.

> Various images scanning tools, such as Clair or Trivy, are useful to audit images from security vulnerabilities. 

Reviewer 2: I would say "Various images scanning tools, including open-source tools like Clair or Trivy, are useful to audit images from security vulnerabilities. ", to inform the reader about open-source tools available.  There are plenty of proprietary tools of course.

Author: Proposal added

-  Report and remove vulnerabilities by upgrading components using authenticated software update distribution portals
-  Adopt a DevSecOps approach and rely on testing automation throughout the software build, integration, delivery, deployment, and runtime operation to perform automatic security check, as described in section 7.4.4 ‘”Infrastructure as a Code Security”

Open-Source Software Security
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Open-source code is present in Cloud Infrastructure software from host Operating System to virtualisation layer 

Reviewer: Also BIOS should be mentioned.

Author: @ulikleber Yes thanks, done

The strength of open-source code is the availability of code source developed by a community which maintain and improve it. 
Open-source code integration with application source code helps to develop and produce applications faster. But, in return, 
it can introduce security risks if a risk management DevSecOps approach is not implemented. The GSMA white paper, “Open 
Networking & the Security of Open Source Software Deployment - Future Networks”, alerts on these risks and addresses the 
challenges coming with open-source code usage. Amongst these risks for security, we can mention a poor quality code 
containing security flaws, an obsolete code with known vulnerabilities, and the lack of knowledge of open source 
communities’ branches activity. An active branch will come with bugs fixes, it will not be the case with an inactive branch. 
The GSMA white paper develops means to mitigate these security issues.

Poor code quality is a factor of risk. Open-source code advantage is its transparency, code can be inspected by tools with 
various capabilities such as open-source software discovery and static and dynamic code analysis.

> A dedicated internal isolated repository separated from the production environment must be used to store vetted open-source 

Reviewer: We may need to explain "internal" here. Each member in the whole chain of software creation must follow this rule (I mean each provider of a component in a software product).

Author: @ulikleber You are right, this principle applies to each actor integrating an open-source component into his software product.
I change the sentence to "Each actor in the whole chain of software production must use a dedicated internal isolated repository separated from the production environment to store..."
